src/main/java/io/debezium/connector/jdbc/JdbcChangeEventSink.java

```diff
@@ -0,0 +1,406 @@
+import io.debezium.connector.jdbc.SinkRecordDescriptor.FieldDescriptor;
+import io.debezium.connector.jdbc.dialect.DatabaseDialect;
+import io.debezium.connector.jdbc.dialect.DatabaseDialectResolver;
+import io.debezium.connector.jdbc.relational.TableDescriptor;
+import io.debezium.data.Envelope;
+import io.debezium.data.Envelope.Operation;
+import io.debezium.pipeline.sink.spi.ChangeEventSink;
+
+/**
+ * A {@link ChangeEventSink} for a JDBC relational database.
+ *
+ * @author Chris Cranford
+ */
+public class JdbcChangeEventSink implements ChangeEventSink {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JdbcChangeEventSink.class);
+
+    private final JdbcSinkConnectorConfig config;
+    private final SessionFactory sessionFactory;
+    private final DatabaseDialect dialect;
+    private final StatelessSession session;
+
+    public JdbcChangeEventSink(JdbcSinkConnectorConfig config) {
+        this.config = config;
+        this.sessionFactory = config.getHibernateConfiguration().buildSessionFactory();
+
+        this.dialect = DatabaseDialectResolver.resolve(config, sessionFactory);
+        this.session = this.sessionFactory.openStatelessSession();
+
+        final DatabaseVersion version = this.dialect.getVersion();
+        LOGGER.info("Database version {}.{}.{}", version.getMajor(), version.getMinor(), version.getMicro());
+    }
+
+    @Override
+    public void execute(SinkRecord record) {
+        try {
+            // Examine the sink record and prepare a descriptor
+            final SinkRecordDescriptor descriptor = SinkRecordDescriptor.builder()
+                    .withPrimaryKeyMode(config.getPrimaryKeyMode())
+                    .withPrimaryKeyFields(config.getPrimaryKeyFields())
+                    .withSinkRecord(record)
+                    .withDialect(dialect)
+                    .build();
+
+            final TableDescriptor table = checkAndApplyTableChangesIfNeeded(record, descriptor);
+            write(table, descriptor);
+        }
+        catch (Exception e) {
+            throw new ConnectException("Failed to process a sink record", e);
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (session != null && session.isOpen()) {
+            LOGGER.info("Closing session.");
+            session.close();
+        }
+        else {
+            LOGGER.info("Session already closed.");
+        }
+
+        if (sessionFactory != null && sessionFactory.isOpen()) {
+            LOGGER.info("Closing the session factory");
+            sessionFactory.close();
+        }
+        else {
+            LOGGER.info("Session factory already closed");
+        }
+    }
+
+    private TableDescriptor checkAndApplyTableChangesIfNeeded(SinkRecord record, SinkRecordDescriptor descriptor) throws SQLException {
+        final String tableName = config.getTableNamingStrategy().resolveTableName(config, record);
+        if (!hasTable(tableName)) {
```

Comment:
Is this cached? I don't think it should be executed for every record as JDBC metadata can be costly operation.